api/valor_api/backend/metrics/text_generation.py

```diff
@@ -0,0 +1,570 @@
+    db: Session,
+    datum_filter: schemas.Filter,
+    groundtruth_filter: schemas.Filter,
+    prediction_filter: schemas.Filter,
+    metrics_to_return: list[MetricType] = [],
+    llm_api_params: dict[str, str | dict] | None = None,
+    metric_params: dict = {},
+) -> Sequence[
+    schemas.AnswerRelevanceMetric
+    | schemas.BLEUMetric
+    | schemas.CoherenceMetric
+    | schemas.ROUGEMetric
+]:
+    """
+    Compute text generation metrics.
+
+    Parameters
+    ----------
+    db : Session
+        The database Session to query against.
+    datum_filter : schemas.Filter
+        The filter to be used to query datums.
+    groundtruth_filter : schemas.Filter
+        The filter to be used to query groundtruths.
+    prediction_filter : schemas.Filter
+        The filter to be used to query predictions.
+    metrics_to_return: list[MetricType]
+        The list of metrics to compute, store, and return to the user.
+    llm_api_params: dict[str, str | dict], optional
+        A dictionary of parameters for the LLM API.
+    metric_params: dict, optional
+        A dictionary of optional parameters to pass in to specific metrics.
+
+    Returns
+    ----------
+    Sequence[schemas.AnswerRelevanceMetric | schemas.BLEUMetric | schemas.CoherenceMetric | schemas.ROUGEMetric]
+        A list of computed metrics.
+    """
+    prediction_subquery = generate_query(
+        models.Annotation.datum_id.label("datum_id"),
+        models.Annotation.text.label("prediction_text"),
+        models.Annotation.context.label("prediction_context"),
+        db=db,
+        label_source=models.Prediction,
+        filters=prediction_filter,
+    ).subquery()
+
+    output = []
+    if any(
+        [metric in TEXT_COMPARISON_METRICS for metric in metrics_to_return]
+    ):
+        # get reference text to compare against from groundtruths
+        # use array_agg since there can be multiple references for a given datum_uid
+        groundtruth_subquery = (
+            generate_query(
+                models.GroundTruth.id,
+                models.Datum.id.label("datum_id"),
+                models.Datum.uid.label("datum_uid"),
+                models.Dataset.name.label("dataset_name"),
+                functions.array_agg(models.Annotation.text).label(
+                    "groundtruth_text"
+                ),
+                db=db,
+                label_source=models.GroundTruth,
+                filters=groundtruth_filter,
+            )
+            .group_by(
+                models.GroundTruth.id,
+                models.Datum.id,
+                models.Datum.uid,
+                models.Dataset.name,
+            )
+            .subquery()
+        )
+
+        joint_subquery = (
+            select(
+                groundtruth_subquery.c.datum_uid,
+                groundtruth_subquery.c.dataset_name,
+                functions.array_agg(
+                    prediction_subquery.c.prediction_text
+                ).label("predictions"),
```

Comment:
@ntlind Why do we array_agg here? If there are multiple predictions for one datum, then array_agg will combine them into a single list. That seems like that could break or at least degrade BLEU and ROUGE performance. It seems like we should be instead returning one BLEU score per prediction, not one BLEU score per datum.

Reply:
Oh I see, I didn't realize that the sentence bleu function below takes in the list of predictions but computes a separate score for each one